localstack/services/sns/provider.py

```diff
@@ -1319,20 +1327,42 @@ def prepare_message_attributes(message_attributes: MessageAttributeMap):
     # todo: Number type is not supported for Lambda subscriptions, passed as String
     #  do conversion here
     for attr_name, attr in message_attributes.items():
-        if attr.get("StringValue", None):
-            val = attr["StringValue"]
-        else:
+        date_type = attr["DataType"]
+        if date_type == "Binary":
             # binary payload in base64 encoded by AWS, UTF-8 for JSON
             # https://docs.aws.amazon.com/sns/latest/api/API_MessageAttributeValue.html
             val = base64.b64encode(attr["BinaryValue"]).decode()
+        else:
+            val = attr.get("StringValue")
 
         attributes[attr_name] = {
-            "Type": attr["DataType"],
+            "Type": date_type,
             "Value": val,
         }
     return attributes
 
 
+def validate_message_attributes(message_attributes: MessageAttributeMap) -> None:
+    """
+    Validate the message attributes, and raises an exception if those do not follow AWS validation
+    :param message_attributes: the message attributes map for the message
+    :raises: InvalidParameterValueException
+    :return: None
+    """
+    for attr_name, attr in message_attributes.items():
+        # `DataType` is a required field for MessageAttributeValue
+        value_key_data_type = "Binary" if (data_type := attr["DataType"]) == "Binary" else "String"
+        value_key = f"{value_key_data_type}Value"
+        if value_key not in attr:
+            raise InvalidParameterValueException(
+                f"The message attribute '{attr_name}' with type '{data_type}' must use field '{value_key_data_type}'."
+            )
```

Comment:
I am not quite sure how it is for SNS, but I think for SQS you could set attributes yourself, and they also would be subject to this check. Which means I could theoretically set DataType=MyNonsenseType, which was not allowed under SQS iirc. This potentially results in the message "The message attribute 'MyAttribute' with type 'MyNonsenseType' must use field 'MyNonsenseTypeValue'
Disregard this if it is handled elsewhere.
Also, for reference: 
https://github.com/localstack/localstack/blob/master/localstack/services/sqs/provider.py#L837
especially
https://github.com/localstack/localstack/blob/master/localstack/services/sqs/provider.py#L860-L864

Reply:
Great catch! I added a test case and the new exception if the DataType is not supported. Thanks a lot for the thorough review 🙏
I added more validation as I was on it, I'll rerequest a quick review to see if nothing is wrong. Thanks for the pointer!